File: SearchSample/Objective-C/SearchSample/ResultViewController.m
//
//  ResultViewController.m
//  SearchSample
//
//  Copyright © 2559 Globtech. All rights reserved.
//

#import "ResultViewController.h"

@implementation ResultViewController


- (void)viewDidLoad {
    [super viewDidLoad];

}

- (void)searchByKeyword:(NSString *)keyword {
    if (keyword != nil) {
        NTLocationSearchParameter *param = [[NTLocationSearchParameter alloc] initWithKeyword:keyword
                                                                                          lat:13.763232
                                                                                          lon:100.538304];

        [self searchWithParam:param];
    }
}

- (void)searchByCategory:(NSString *)category {
    if (category != nil ) {
        NTLocationSearchParameter *param = [[NTLocationSearchParameter alloc] initWithCategoryCode:@[category]
                                                                                               lat:13.763232
                                                                                               lon:100.538304];
        [self searchWithParam:param];
    }
}

- (void)searchByLocalCategory:(NSString *)localCategory {
    if (localCategory != nil ) {
        NTLocationSearchParameter *param = [[NTLocationSearchParameter alloc] initWithLocalCategoryCode:@[localCategory]
                                                                                                    lat:13.763232
                                                                                                    lon:100.538304];

        [self searchWithParam:param];
    }
}

- (void)searchWithParam:(NTLocationSearchParameter *)param {
    [NTLocationSearchService executeAsync:param Completion:^(NTLocationSearchResultSet * resultSet, NSError * error) {
        if (!error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                _result = resultSet.results;
                [_tableView reloadData];
            });
        }
        else {
            NSLog(@"error: %@",error.localizedDescription);
        }
    }];
}

- (void)performUnabletoSearch {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Unable to search" message:@"Please check your location." preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"OK"
                                                        style:UIAlertActionStyleCancel
                                                      handler:^(UIAlertAction * _Nonnull action) {
        [self.navigationController popViewControllerAnimated:true];
    }]];
    
    [self presentViewController:alertController animated:true completion:nil];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"mapResultSegue"]) {
        MapResultViewController *mapResultViewController = segue.destinationViewController;
        mapResultViewController.result = sender;
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NTLocationSearchResult *result = _result[indexPath.row];
    
    [self performSegueWithIdentifier:@"mapResultSegue" sender:result];
    
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    
    NTLocationSearchResult *result = _result[indexPath.row];
    
    cell.textLabel.text = result.name_L;
    cell.detailTextLabel.text = [NSString stringWithFormat:@"%@, %@, %@", result.adminLevel3_L, result.adminLevel2_L, result.adminLevel1_L];
    
    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _result != nil ? _result.count : 0;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}


@end

File: SearchSample/Swift/SearchSample/ResultViewController.swift
//
//  ResultViewController.swift
//  SearchSample
//
//  Copyright © 2559 Globtech. All rights reserved.
//

import UIKit
import NOSTRASDK
import CoreLocation

class ResultViewController: UIViewController {

    @IBOutlet weak var tableView: UITableView!
    
    
    var results: [NTLocationSearchResult]?;
    let locationManager = CLLocationManager()
    var coordinate: CLLocationCoordinate2D! = nil;
    override func viewDidLoad() {
        super.viewDidLoad();
        
        // For use in foreground
        coordinate = self.locationManager.location?.coordinate;
        
    }
    
    func searchByKeyword(_ keyword: String!) {
        if keyword != nil && coordinate != nil {
            let param = NTLocationSearchParameter(keyword: keyword,
                                                  lat: (coordinate?.latitude)!,
                                                  lon: (coordinate?.longitude)!);
            self.searchWithParam(param);
        }
        else {
            self.performUnabletoSearch();
        }
    }
    
    func searchByCategory(_ category: String!) {
        if category != nil && coordinate != nil {
            
            let param = NTLocationSearchParameter(categoryCode: [category],
                                                  lat: (coordinate?.latitude)!,
                                                  lon: (coordinate?.longitude)!);
            

            self.searchWithParam(param);
        }
        else {
            self.performUnabletoSearch();
        }

    }
    
    func searchByLocalCategory(_ localCategory: String!) {
        if localCategory != nil && coordinate != nil {
            
            
            let param = NTLocationSearchParameter(localCategoryCode: [localCategory],
                                                  lat: (coordinate?.latitude)!,
                                                  lon: (coordinate?.longitude)!);
            
            
            self.searchWithParam(param);
        }
        else {
            self.performUnabletoSearch();
        }
    }
    
    
    func searchWithParam(_ param: NTLocationSearchParameter) {
        NTLocationSearchService.executeAsync(param, Completion: { (resultSet, error) in
            DispatchQueue.main.async(execute: {
                if error == nil {
                    self.results = resultSet?.results;
                }
                else {
                    self.results = [];
                    print("error: \(error?.description)");
                }
                self.tableView.reloadData();
            })
            
        });
    }
    
    
    func performUnabletoSearch() {
        let alertController = UIAlertController(title: "Unable to search", message: "Please check your location.", preferredStyle: .alert);
        alertController.addAction(UIAlertAction(title: "OK", style: .cancel, handler: { (action) in
            _ = self.navigationController?.popViewController(animated: true);
        }));
        
        self.present(alertController, animated: true, completion: nil);
    }
    
    
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "mapResultSegue" {
            let mapResultViewController = segue.destination as? MapResultViewController;
            mapResultViewController?.result = sender as! NTLocationSearchResult;
        }
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAtIndexPath indexPath: IndexPath) {
        let result = results![indexPath.row];
        self.performSegue(withIdentifier: "mapResultSegue", sender: result);
    }
    
    
    func tableView(_ tableView: UITableView, cellForRowAtIndexPath indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell");
        let result = results![indexPath.row];
        
        cell?.textLabel?.text = result.name_L;
        cell?.detailTextLabel?.text = "\(result.adminLevel3_L), \(result.adminLevel2_L), \(result.adminLevel1_L)"
        
        return cell!;
        
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return results != nil ? (results?.count)! : 0;
    }
    
    func numberOfSectionsInTableView(_ tableView: UITableView) -> Int {
        return 1
    }
    
    
}
